// tslint/build/rules/banRule.js
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        _super.apply(this, arguments);
    }
    Rule.prototype.apply = function (syntaxTree) {
        var options = this.getOptions();
        var banFunctionWalker = new BanFunctionWalker(syntaxTree, options);
        var functionsToBan = options.ruleArguments;
        functionsToBan.forEach(function (functionToBan) {
            banFunctionWalker.addBannedFunction(functionToBan);
        });
        return this.applyWithWalker(banFunctionWalker);
    };
    Rule.FAILURE_STRING_PART = "function invocation disallowed: ";
    return Rule;
})(Lint.Rules.AbstractRule);
exports.Rule = Rule;

var BanFunctionWalker = (function (_super) {
    __extends(BanFunctionWalker, _super);
    function BanFunctionWalker() {
        _super.apply(this, arguments);
        this.bannedFunctions = [];
    }
    BanFunctionWalker.prototype.addBannedFunction = function (bannedFunction) {
        this.bannedFunctions.push(bannedFunction);
    };

    BanFunctionWalker.prototype.visitInvocationExpression = function (node) {
        var _this = this;
        var expression = node.expression;

        if (expression.kind() === 212 /* MemberAccessExpression */ && expression.childCount() >= 3) {
            var firstToken = expression.firstToken();
            var secondToken = expression.childAt(1);
            var thirdToken = expression.childAt(2);

            if (secondToken.kind() === 76 /* DotToken */) {
                this.bannedFunctions.forEach(function (bannedFunction) {
                    if (firstToken.text() === bannedFunction[0] && thirdToken.fullText() === bannedFunction[1]) {
                        var position = _this.position() + node.leadingTriviaWidth();
                        _this.addFailure(_this.createFailure(position, expression.width(), Rule.FAILURE_STRING_PART + firstToken.text() + "." + thirdToken.fullText()));
                    }
                });
            }
        }

        _super.prototype.visitInvocationExpression.call(this, node);
    };
    return BanFunctionWalker;
})(Lint.RuleWalker);
exports.BanFunctionWalker = BanFunctionWalker;
